core/ir/riscv64/codec.c

```diff
@@ -0,0 +1,1066 @@
+        else if (rv64)
+            return &instr_infos[OP_c_addiw];
+        else
+            return NULL;
+    case 2: return &instr_infos[OP_c_li];
+    case 3:
+        if (bit11to7 == 2)
+            return &instr_infos[OP_c_addi16sp];
+        else
+            return &instr_infos[OP_c_lui];
+    case 4:
+        switch (funct2) {
+        case 0: return &instr_infos[OP_c_srli];
+        case 1: return &instr_infos[OP_c_srai];
+        case 2: return &instr_infos[OP_c_andi];
+        case 3:
+            switch (bit12) {
+            case 0:
+                switch (funct3) {
+                case 0: return &instr_infos[OP_c_sub];
+                case 1: return &instr_infos[OP_c_xor];
+                case 2: return &instr_infos[OP_c_or];
+                case 3: return &instr_infos[OP_c_and];
+                default: return NULL;
+                }
+            case 1:
+                switch (funct3) {
+                case 0: return &instr_infos[OP_c_subw];
+                case 1: return &instr_infos[OP_c_addw];
+                // 2 and 3 are reserved.
+                default: return NULL;
+                }
+            default: return NULL;
+            }
+        default: return NULL;
+        }
+    case 5: return &instr_infos[OP_c_j];
+    case 6: return &instr_infos[OP_c_beqz];
+    case 7: return &instr_infos[OP_c_bnez];
+    default: return NULL;
+    };
+}
+
+static inline rv_instr_info_t *
+match_op_2(int funct, int bit11to7, int bit6to2, int bit12, bool rv32, bool rv64)
+{
+    switch (funct) {
+    case 0: return &instr_infos[OP_c_slli];
+    case 1: return &instr_infos[OP_c_fldsp];
+    case 2: return &instr_infos[OP_c_lwsp];
+    case 3:
+        if (rv32)
+            return &instr_infos[OP_c_flwsp];
+        else if (rv64)
+            return &instr_infos[OP_c_ldsp];
+        else
+            return NULL;
+    case 4:
+        switch (bit12) {
+        case 0:
+            if (bit6to2 == 0)
+                return &instr_infos[OP_c_jr];
+            else
+                return &instr_infos[OP_c_mv];
+        case 1:
+            if ((bit11to7 == 0) && (bit6to2 == 0))
+                return &instr_infos[OP_c_ebreak];
+            else if (bit6to2 == 0)
+                return &instr_infos[OP_c_jalr];
+            else
+                return &instr_infos[OP_c_add];
+        default: return NULL;
+        }
+    case 5: return &instr_infos[OP_c_fsdsp];
+    case 6: return &instr_infos[OP_c_swsp];
+    case 7:
+        if (rv32)
+            return &instr_infos[OP_c_fswsp];
+        else if (rv64)
+            return &instr_infos[OP_c_sdsp];
+        else
+            return NULL;
+    default: return NULL;
+    }
+}
+
+static rv_instr_info_t *
+get_rvc_instr_info(uint32_t inst, int xlen)
+{
+    /* 0 is an illegal instruction which is often used as a canary */
+    if (inst == 0)
+        return &instr_infos[OP_unimp];
+
+    int op = inst & 0b11;
+    int funct = (inst >> 13) & 0b111;
```

Comment:
Could we use GET_FIELD for all of these?

Reply:
Yes and I should have done that.

Reply:
Done